### ISSUE TITLE
Installation issue: boost%nvhpc

### ISSUE THREAD

**Issue by pjh40** [LLNL/spack#24489]
### Steps to reproduce the issue

```console
$ spack install boost@1.76.0%nvhpc@21.5
...
==> Installing boost-1.76.0-egxydhpfoggdbo6iob76xudt3sysrxcl
==> No binary for boost-1.76.0-egxydhpfoggdbo6iob76xudt3sysrxcl found: installing from source
==> Using cached archive: /spack/var/spack/cache/_source-cache/archive/f0/f0397ba6e982c4450f27bf32a2a83292aba035b827a5623a14636ea583318c41.tar.bz2
3 out of 3 hunks FAILED -- saving rejects to file tools/build/src/engine/build.sh.rej
==> Patch /spack/var/spack/repos/builtin/packages/boost/nvhpc.patch failed.
==> Error: ProcessError: Command exited with status 1:
    '/usr/bin/patch' '-s' '-p' '1' '-i' '/spack/var/spack/repos/builtin/packages/boost/nvhpc.patch' '-d' '.'
```
It appears that the `nvhpc.patch` added by @samcmill to override the PGI configuration in  `tools/build/src/engine/build.sh` with nvhpc compilers is out of date. It would be nice if this configuration was supported natively.

### Information on your system

* **Spack:** 0.16.2-3241-1bccd866ae
* **Python:** 3.6.8
* **Platform:** linux-rhel7-power9le
* **Concretizer:** original

### Additional information
* [build.sh.rej.txt](https://github.com/spack/spack/files/6703622/build.sh.rej.txt)

@hainest 
### General information

- [x] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [x] I have run `spack maintainers <name-of-the-package>` and @mentioned any maintainers
- [x] I have uploaded the build log and environment files
- [x] I have searched the issues of this repo and believe this is not a duplicate

**Comment by pjh40** [comment]
Note that the existing patch works for `boost@1.75.0`, so this failure is due to a change introduced at `boost@1.76.0`.

**Comment by pjh40** [comment]
For `nvhpc@21.5` at least, it appears that boost is picking up the correct compiler *without* the patch applied at all.  From the `spack-build-01-install-out.txt` file:
```console
==> boost: Executing phase: 'install'
==> [2021-06-23-12:04:59.552592] './bootstrap.sh' '--prefix=/spack/opt/spack/linux-rhel7-power9le/nvhpc-21.5/boost-1.76.0-2k7wdcusmquscgyipytvm32kkgfea6zn' '--with-toolset=pgi' '--with-libraries=iostreams,filesystem,system,locale,exception,math,graph,thread,wave,serialization,atomic,timer,chrono,test,random,regex,log,program_options,date_time' '--without-icu'
Building B2 engine..

###
###
### Using 'pgi' toolset.
###
###


pgc++ (aka nvc++) 21.5-0 linuxpower target on Linuxpower 
PGI Compilers and Tools
Copyright (c) 2021, NVIDIA CORPORATION.  All rights reserved.

###
###
```
and `spack install` was successful.  Perhaps this issue could just be resolved with:
```diff
diff --git a/var/spack/repos/builtin/packages/boost/package.py b/var/spack/repos/builtin/packages/boost/package.py
index a0c3656941..63fe5a7bd5 100644
--- a/var/spack/repos/builtin/packages/boost/package.py
+++ b/var/spack/repos/builtin/packages/boost/package.py
@@ -217,7 +217,7 @@ def libs(self):
     patch('boost_1.63.0_pgi_17.4_workaround.patch', when='@1.63.0%pgi@17.4')
 
     # Patch to override the PGI toolset when using the NVIDIA compilers
-    patch('nvhpc.patch', when='%nvhpc')
+    patch('nvhpc.patch', when='@1.74.0:1.75.9999%nvhpc')
 
     # Fix for version comparison on newer Clang on darwin
     # See: https://github.com/boostorg/build/issues/440
```
but I'm not sure about the `tools/build/src/tools/pgi.jam` part of the patch, or if this works for earlier versions of `nvhpc`.

**Comment by hainest** [comment]
>For nvhpc@21.5 at least, it appears that boost is picking up the correct compiler without the patch applied at all.

I'm assuming that's for boost@1.76.0?

I'm not familiar with PGI, so I'm going to have to defer to @samcmill for some ideas.

**Comment by pjh40** [comment]
Yes, it is for boost@1.76.0

**Comment by samcmill** [comment]
Based on the above, the patch should definitely be made conditional on the boost version as shown in https://github.com/spack/spack/issues/24489#issuecomment-867058888.  

I will need to investigate further to see whether an additional patch is still needed for 1.76.0 under some circumstances or not.

**Comment by pjh40** [comment]
In a more complicated build environment, it appears that I'm now getting an ICE on 1.75.0 (with your patch) *and* 1.76.0 (no patch) with nvhpc@21.5. Same location for both boost versions. I'll investigate more tomorrow and report back.

**Comment by samcmill** [comment]
1.74.0 builds for me with 21.3 and 21.5.  

I'm getting `NVC++-F-0000-Internal compiler error. msz_dtype, bad value     440  (libs/atomic/src/find_address_sse2.cpp: 60)` for 1.75.0 and 1.76.0.  I can investigate more later.

**Comment by pjh40** [comment]
My errors looked like:
```console
2 errors found in build log:
     528      pgi.compile.c++ bin.v2/libs/mpi/build/pgi/release/cxxstd-14-iso/l
              ink-static/error_string.o
     529      pgi.compile.c++ bin.v2/libs/log/build/pgi/release/cxxstd-14-iso/l
              ink-static/threadapi-pthread/threading-multi/syslog_backend.o
     530      "libs/log/src/syslog_backend.cpp", line 166: warning: variable "l
              ock" was declared but never referenced
     531                    lock_guard< mutex > lock(mutex_holder::get());
     532                                        ^
     533      
  >> 534      "./boost/asio/execution/context_as.hpp", line 145: internal error
              : assertion failed at: "edg_error.cpp", line 1300 in form_templat
              e_arg_info
     535      
     536        const U context_as_t<T>::static_query_v;
     537              ^
     538      
     539      1 error and 1 catastrophic error detected in the compilation of "
              libs/log/src/syslog_backend.cpp".
     540      Compilation aborted.

     ...

     16493    pgi.compile.c++ bin.v2/libs/math/build/pgi/release/cxxstd-14-iso/
              fmin.o
     16494    pgi.compile.c++ bin.v2/libs/log/build/pgi/release/cxxstd-14-iso/t
              hreadapi-pthread/threading-multi/syslog_backend.o
     16495    "libs/log/src/syslog_backend.cpp", line 166: warning: variable "l
              ock" was declared but never referenced
     16496                  lock_guard< mutex > lock(mutex_holder::get());
     16497                                      ^
     16498    
  >> 16499    "./boost/asio/execution/context_as.hpp", line 145: internal error
              : assertion failed at: "edg_error.cpp", line 1300 in form_templat
              e_arg_info
     16500    
     16501      const U context_as_t<T>::static_query_v;
     16502            ^
     16503    
     16504    1 error and 1 catastrophic error detected in the compilation of "
              libs/log/src/syslog_backend.cpp".
     16505    Compilation aborted.
```
The full spec for the boost that I was building is:
```
boost@1.76.0%nvhpc@21.5+atomic+chrono~clanglibcpp~container\
    ~context~coroutine+date_time~debug+exception~fiber+filesystem+graph~icu+iostreams\
    +locale+log+math+mpi+multithreaded~numpy+pic+program_options~python+random+regex\
    +serialization+shared+signals~singlethreaded+system~taggedlayout+test+thread\
    +timer~versionedlayout+wave cxxstd=14 visibility=hidden arch=linux-rhel7-power9le
```
I'll try using 1.74.0 for now.

**Comment by pjh40** [comment]
Interesting... that nasty spec above was basically specified by `boost%nvhpc+mpi+pic cxxstd=14`.  When I drop the `cxxstd=14` variant, 1.76.0 builds find with nvhpc@21.5 and nvhpc mpi, which is good!

**Comment by samcmill** [comment]
I've filed https://github.com/spack/spack/pull/24541 to fix this.  

Besides the build environment issues, this also surfaced a compiler bug when building https://github.com/boostorg/atomic/blob/a548b3c15c97d9c1b89e0bb16518baeaeb7a9aa3/src/find_address_sse2.cpp#L55.  That code was introduced in Boost 1.75.0.

**Comment by hainest** [comment]
Thanks, @samcmill! You may want to add a 'closes #24489' to auto-close this issue when that PR is merged.

**Comment by pjh40** [comment]
Works for me, thanks for the fix!
